File: ListNode/CopyList.cpp
#include <iostream>
#include <unordered_map>

using std::unordered_map;

class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node() {}

    Node(int _val, Node* _next, Node* _random) {
        int val = _val;
        Node* next = _next;
        Node* random = _random;
    }
};

class S {
public:
    Node* copyRandomList(Node* head) {
        Node dummy(-1, head, NULL);
        return dummy.next;
    }
};

int main()
{
    unordered_map<Node*, Node*> mymap;
    Node dummy(5, NULL, NULL);
    Node* p = &dummy;

    for (int i = 1; i < 5; i++) {
        p->next = new Node;
        p->next->val = i;
        p->next->next = NULL;
        p->next->random = NULL;
        p = p->next;
    }
    Node* q = dummy.next;
    while (q) {
        std::cout << q->val << "\n";
        q = q->next;
    }

    // mymap[p] = new Node(p->val, NULL, NULL);

    return 0;
}

File: LinkedList/CopyList.py

# Definition for a Node.

class Node:
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':

        if not head:
            return head

        map = {}

        current = head
        while current:
            map[current] = Node(current.val, None, None)
            current = current.next

        current = head
        while current:
            if current.next:
                map[current].next = map[current.next]
            else:
                map[current].next = None

            if current.random:
                map[current].random = map[current.random]
            else:
                map[current].random = None

            current = current.next

        return map[head]


# class Solution:
#     def copyRandomList(self, head: 'Node') -> 'Node':
#         import copy
#         return copy.deepcopy(head)
